test/e2e/conformance/webhook_test.go

```diff
@@ -0,0 +1,224 @@
+		framework.NewWorkspaceFixture(t, server, organization, "Universal"),
+		framework.NewWorkspaceFixture(t, server, organization, "Universal"),
+	}
+	clientsets, err := kubernetes.NewClusterForConfig(cfg)
+	cowbyClients, err := client.NewClusterForConfig(cfg)
+	require.NoError(t, err, "failed to construct apiextensions client for server")
+	//Insall the Cowboy resources and the webhook in the first logical cluster
+	for i, logicalCluster := range logicalClusters {
+		t.Logf("Bootstrapping ClusterWorkspace CRDs in logical cluster %s", logicalCluster)
+		apiExtensionsClients, err := apiextensionsclient.NewClusterForConfig(cfg)
+		require.NoError(t, err, "failed to construct apiextensions client for server")
+		crdClient := apiExtensionsClients.Cluster(logicalCluster).ApiextensionsV1().CustomResourceDefinitions()
+		wildwest.Create(t, crdClient, metav1.GroupResource{Group: "wildwest.dev", Resource: "cowboys"})
+
+		// install the webhook in the first logical cluster base level.
+		url := "https://localhost:8090/hello"
+		sideEffect := admissionregistrationv1.SideEffectClassNone
+		if i == 0 {
+			webhook := &admissionregistrationv1.ValidatingWebhookConfiguration{
+				TypeMeta:   metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{Name: "test-webhook"},
+				Webhooks: []admissionregistrationv1.ValidatingWebhook{{
+					Name: "test-webhook.cowboy.io",
+					ClientConfig: admissionregistrationv1.WebhookClientConfig{
+						URL:      &url,
+						CABundle: cfg.CAData,
+					},
+					Rules: []admissionregistrationv1.RuleWithOperations{{
+						Operations: []admissionregistrationv1.OperationType{
+							admissionregistrationv1.Create,
+						},
+						Rule: admissionregistrationv1.Rule{
+							APIGroups:   []string{"wildwest.dev"},
+							APIVersions: []string{"v1alpha1"},
+							Resources:   []string{"cowboys"},
+						},
+					}},
+					SideEffects:             &sideEffect,
+					AdmissionReviewVersions: []string{"v1"},
+				}},
+			}
+			_, err := clientsets.Cluster(logicalCluster).AdmissionregistrationV1().ValidatingWebhookConfigurations().Create(ctx, webhook, metav1.CreateOptions{})
+			require.NoError(t, err, "failed to construct apiextensions client for server")
+		}
+	}
+	cowboy := v1alpha1.Cowboy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "testing",
+			Namespace: "default",
+		},
+		Spec: v1alpha1.CowboySpec{},
+	}
+	_, err = cowbyClients.Cluster(logicalClusters[1]).WildwestV1alpha1().Cowboys("default").Create(ctx, &cowboy, metav1.CreateOptions{})
+	require.NoError(t, err, "failed to create cowboy resource in second logical cluster")
+	if testWebhook.Calls != 0 {
+		t.Fail()
+	}
+	_, err = cowbyClients.Cluster(logicalClusters[0]).WildwestV1alpha1().Cowboys("default").Create(ctx, &cowboy, metav1.CreateOptions{})
+	require.NoError(t, err, "failed to create cowboy resource in second logical cluster")
+	if testWebhook.Calls != 1 {
+		t.Fatalf("Calls is %v but expected %v", testWebhook.Calls, 1)
+	}
```

Comment:
test in general looks fine. You will notice that you run into some race:
1. the CRD creation and the server being ready to serve are asynchronous, you your Create here will possibly fail for O(tens or hundreds of milliseconds). `require.Eventually` will help.
2. the admission webhook is informer driven inside of kube. That's also asynchronously update. I.e. it could be that these Create work, but your webhook is not yet called. Again, some `require.Eventually` will help.

Reply:
Thank you for this knowledge! 

If I understand correctly I need to eventually require that the create call will work. 

I would like to understand point 2 more fully for my knowledge. Does this mean that I could expect that a controller gets an object that did not get processed by the validating webhook?

Thanks.